packages/nx/src/project-graph/affected/locators/npm-packages.ts

```diff
@@ -31,6 +31,7 @@ export const getTouchedNpmPackages: TouchedProjectLocator<
         const npmPackage = npmPackages.find(
           (pkg) => pkg.data.packageName === c.path[1]
         );
+        if (!npmPackage) continue;
```

Comment:
This shouldn't happen. All packages should be found here. Every change  to `package.json#dependencies`  and `package.json#devDependencies` should cause a project to be touched.

But throwing an error might be too extreme.

Let's print a warning instead.
1. Let's only print 1 warning for this whole process though (so we don't get a warning for each package that wasn't found.)
2. The warning should include the packages that weren't found.
3. The warning should tell the user to open an issue in Github including the `package.json` file.  

This way people can still run `affected` but we can identify cases we haven't handled and fix them.

cc @meeroslav

Reply:
Thanks for taking the time to review this @FrozenPandaz and for the suggestions.

> This shouldn't happen. All packages should be found here.

To give context as to how this happened to my project. A new workspace project (with a package.json that was intended to be published to a registry) was created and is used within the existing `nx` workspace that already is using `affected`. The new workspace project was then added in the workspace as a pnpm `workspace:*` dependency.

I have made the suggested changes, do let me know if there is anything I didn't capture in the last commit https://github.com/nrwl/nx/pull/13681/commits/221fb6635623587f81e7ded34b7d631ca0ec057d

Reply:
This issue should not be hidden under the carpet. The problem is that we are expecting a workspace project to exist in the `externalNodes` but it does not. It used to, though, since previously we didn't parse lock files so everything in the `package.json` dependencies block was treated as an external dependency.

Instead, if the package is not found, we need to check `projectGraph.nodes` if it can be found there.

Having a fallback with a warning is fine, but let's exhaust all the possibilities first.

Reply:
I believe this is the correct way to fix it: https://github.com/nrwl/nx/pull/13728

@FrozenPandaz do you think we should still keep this fallback? I guess this can easily happen when the lock file (install has not been run) or file system (dependency does not match the workspace project name) and package.json are not in sync.